site/_posts/2020-05-22-building-cli-app-with-device-grant-and-golang.md

```diff
@@ -0,0 +1,303 @@
+		faClient = fusionauth.NewClient(httpClient, baseURL, apiKey)
+		openIDConfig, err := faClient.RetrieveOpenIdConfiguration()
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		deviceResp, err := startDeviceGrantFlow(openIDConfig.DeviceAuthorizationEndpoint)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		informUserAndOpenBrowser(deviceResp.UserCode)
+
+		accessToken, err := startPolling(openIDConfig.TokenEndpoint, deviceResp.DeviceCode, deviceResp.Interval)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+      // To be continued...
+   },
+}
+
+// ...
+
+func informUserAndOpenBrowser(userCode string) {
+	cyan := color.New(color.FgCyan)
+	cyan.Printf("Your User Code is: ")
+
+	yellow := color.New(color.FgYellow, color.Bold)
+	yellow.Printf("%s\n", userCode)
+
+	cyan.Printf("Opening browser for code entry...\n")
+
+	// Wait a few seconds to give the user a chance to check out the printed user code.
+	time.Sleep(3 * time.Second)
+
+	url := fmt.Sprintf("%s/oauth2/device?client_id=%s&tenantId=%s", host, ClientID, TenantID)
+	open.Run(url)
+}
+
+func startPolling(tokenEndpoint string, deviceCode string, retryInterval int) (*fusionauth.AccessToken, error) {
+	var result *fusionauth.AccessToken = &fusionauth.AccessToken{}
+	yellow := color.New(color.FgYellow, color.Bold)
+	blue := color.New(color.FgBlue, color.Bold)
+
+	for {
+		resp, err := http.PostForm(tokenEndpoint, url.Values{
+			"device_code": {deviceCode},
+			"grant_type":  {"urn:ietf:params:oauth:grant-type:device_code"},
+			"client_id":   {ClientID},
+		})
+
+		if err != nil {
+			return result, err
+		}
+
+		// 400 status code (StatusBadRequest) is our sign that the user
+		// hasn't completed their device login yet, sleep and then continue.
+		if resp.StatusCode == http.StatusBadRequest {
+
+			// Sleep for the retry interval and print a dot for each second.
+			for i := 0; i < retryInterval; i++ {
+				if i == 0 {
+					blue.Printf(".")
+				} else {
+					yellow.Printf(".")
+				}
+				time.Sleep(time.Second)
+			}
+
+			continue
+		}
+
+		fmt.Printf("\n")
+		defer resp.Body.Close()
+		json.NewDecoder(resp.Body).Decode(result)
+
+		return result, nil
+	}
+}
+
+// ...
+```
+
+This should look familiar! We start with our LoginCmd's `Run` function again, but we continue that function a bit more to do the following:
+
+1. We call `informUserAndOpenBrowser` with our `deviceResp.UserCode` that we got from the FA server. This prints to the user's CLI the given `userCode` and open's up FA's `/oauth2/device` endpoint for our Application and Tenant.
+1. Next, we call `startPolling` which continually polls the given `openIDConfig.TokenEndpoint` every `retryInterval` seconds (5 seconds is RFC8628's / FA's default) to see if the user has completed the Device login (tied to the given `deviceCode`). Once the user has completed their login, the FA server responds with a proper `200` response including a JWT Access Token which then populates our returned `result` object.
```

Comment:
Not sure which style guide we are using, but I believe some of them require commas before "which".